### PR TITLE
Initcode transaction (TXCREATE) support

### PR DESCRIPTION
**Pull request by gumb0**
Requires https://github.com/ethereum/evmc/pull/681

### CODE REVIEW

File: include/evmc/evmc.h [hunk 1]
```diff
@@ -880,6 +901,9 @@ struct evmc_host_interface
 
     /** Set transient storage callback function. */
     evmc_set_transient_storage_fn set_transient_storage;
+
+    /** Get transaction initcode by hash callback function. */
+    evmc_get_tx_initcode_by_hash_fn get_tx_initcode_by_hash;
```
Comment (gumb0) [inline include/evmc/evmc.h:906]: Does it make sense to put it close to `get_tx_context` ?

Comment (pdobacz) [reply include/evmc/evmc.h:906]: as in, move it up/down in the file? I think so. I can't remember exactly why I positioned stuff where I did anymore.

File: examples/example_host.cpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 /// @file
 /// Example implementation of an EVMC Host.
+/// TODO: Doesn't support TXCREATE transaction initcodes.
```
Comment (gumb0) [inline examples/example_host.cpp:7]: Not sure it makes sense to support it, `example.c` only sends `EVMC_CALL` message anyway.

Comment (pdobacz) [reply examples/example_host.cpp:7]: feel free to drop the comment like this, if you feel like nothing needs to be done. I was very generous when adding them in

**Review by pdobacz — APPROVED** [review]

**Review by chfast — COMMENTED** [review]
I don't want to spoil the party, but I think the simpler API would be to extend the `evmc_tx_context` with

```cpp
hashed_initcode* initcodes;
size_t initcodes_count;
```

where 

```cpp
struct hashed_initcode {
    evmc_bytes32 hash;
    const uint8_t* code;
    size_t code_size;
};
```

**Comment by pdobacz** [comment]
> I don't want to spoil the party, but I think the simpler API would be to extend the `evmc_tx_context` with

Indeed simpler but makes problematic:
1. Optimizing initcode lookup (unless we require the initcodes are sorted I suppose)
2. Any usecases where the initcode is fetched not from the TX, for which the fetch-by-hash was introduced in the first place

Neither is maybe big deal but sth to keep in mind.

**Comment by gumb0** [comment]
> Indeed simpler but makes problematic:
> 
> 1. Optimizing initcode lookup (unless we require the initcodes are sorted I suppose)

I don't think this variant changes much, it only changes where lookup by hash is happening.
-  With current API, Host gets initcodes array from transaction and hashes them and puts them into `unordered_map` to later serve `get_tx_initcode_by_hash()` requests. 
- With this new API, Host will get initcode array from transaction, hash them once and serve as array in `tx_context`, then EVM will get them and put into `unordered_map` on EVM side (in evmone it will be in `ExecutionState`).

> 2. Any usecases where the initcode is fetched not from the TX, for which the fetch-by-hash was introduced in the first place

For something other than transaction, it will be some other different API? But I think fetch-by-hash is mainly for easier aggregating several transactions (or AA-transactions) into one batch transaction (merging their initcodes without changing TXCREATE instances).

**Comment by gumb0** [comment]
Anyway I plan to try this in a different PR.

**Comment by pdobacz** [comment]
Makes sense!

**Comment by chfast** [comment]
The other approach can be implemented as a "flat map": vector of already sorted (hash, initcode) pairs. EVM only need to read it and can lookup values by binary search.

But I'm not insisting on sorting values right now. Linear search should be fine as the start.

**Comment by gumb0** [comment]
> The other approach can be implemented as a "flat map": vector of already sorted (hash, initcode) pairs. EVM only need to read it and can lookup values by binary search.

https://github.com/ethereum/evmone/pull/702/commits/55a7102c2f4f13ccf372b597ad4b3f032002ef86
EVM puts them into `unordered_map`.

**Comment by gumb0** [comment]
Replaced by #709

**Pull request closed without merging**